File: sourcecode/models/task.py
import enum

from sqlalchemy import (
    Column,
    Integer,
    String,
    Text,
    Date,
    DateTime,
    Enum,
    ForeignKey,
    func,
)
from sqlalchemy.orm import relationship

from sourcecode.db.database import Base


class PriorityEnum(str, enum.Enum):
    High = "High"
    Medium = "Medium"
    Low = "Low"


class StatusEnum(str, enum.Enum):
    todo = "To Do"
    in_progress = "In Progress"
    done = "Done"


class Note(Base):
    __tablename__ = "notes"

    id = Column(Integer, primary_key=True, autoincrement=True)
    raw_text = Column(Text, nullable=False)
    created_at = Column(DateTime, server_default=func.now())

    tasks = relationship("Task", back_populates="note")


class Task(Base):
    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, autoincrement=True)
    note_id = Column(Integer, ForeignKey("notes.id"), nullable=True)
    description = Column(Text, nullable=False)
    due_date = Column(Date, nullable=True)
    owner = Column(String(100), nullable=True)
    priority = Column(Enum(PriorityEnum), nullable=False, default=PriorityEnum.Medium)
    status = Column(Enum(StatusEnum), nullable=False, default=StatusEnum.todo)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    note = relationship("Note", back_populates="tasks")

File: client/src/types/task.ts
export type TaskStatus = 'To Do' | 'In Progress' | 'Done';
export type TaskPriority = 'High' | 'Medium' | 'Low';

export interface Task {
  id: number;
  description: string;
  owner?: string;
  priority: TaskPriority;
  status: TaskStatus;
  dueDate?: string;
  createdAt: string;
  updatedAt: string;
}

// ---------- Extraction types ----------

/** A new task from the LLM — not yet saved. Shown in the confirmation modal. */
export interface ProposedNewTask {
  description: string;
  dueDate?: string;      // missing = unstructured
  owner?: string;        // missing = unstructured
  priority: TaskPriority;
}

/** A proposed change to an existing saved task (PM must approve before it is applied). */
export interface ProposedUpdate {
  taskId: number;
  description: string;
  current: Record<string, unknown>;
  changes: Record<string, unknown>;
}

/** Shape returned by POST /api/extract — nothing is saved yet. */
export interface ExtractionResponse {
  noteText: string;
  structuredNew: ProposedNewTask[];    // have due_date AND owner
  unstructuredNew: ProposedNewTask[];  // missing due_date or owner
  proposedUpdates: ProposedUpdate[];   // changes to existing tasks
}

/** Shape returned by POST /api/extract/confirm-all */
export interface ConfirmAllResponse {
  noteId: number;   // ID of the saved note — passed back to confirmUpdates for audit linking
  created: Task[];
  updated: Task[];
}

// ---------- Task notes history ----------

/** A note that created or updated a task — returned by GET /api/tasks/{id}/notes. */
export interface NoteLink {
  noteId: number;
  noteText: string;
  linkType: 'created' | 'updated';
  createdAt: string;
}

// ---------- Notes ----------

export interface NoteEntry {
  id: number;
  rawText: string;
  createdAt: string;
  taskCount: number;
}

// ---------- Draft / save state ----------

/** In-memory per-row unsaved changes. Keys match Task field names. */
export type TaskDraft = Partial<Pick<Task, 'owner' | 'dueDate' | 'priority' | 'status'>>;

// ---------- Sorting ----------

export type SortDirection = 'asc' | 'desc';

export interface SortState<T> {
  key: keyof T | null;
  direction: SortDirection;
}
